### PR TITLE
Add PostgreSQL FeatureFlagStore

### PR DESCRIPTION
**Pull request by adambom**
Fixes: https://github.com/carta/flipper-client/issues/44

Author: @gpouilloux. For some reason the permissions got messed up so he is unable to create a PR from the fork.

### CODE REVIEW

File: flipper/contrib/postgresql.py
Comment (adambom) [inline flipper/contrib/postgresql.py:78]: How would you feel about condensing this down into a single query? Since you're using the feature name as the pk, there should automatically be a unique constraint. This would allow us to use `INSERT ... ON CONFLICT DO NOTHING` semantics.

File: flipper/contrib/postgresql.py
Comment (adambom) [inline flipper/contrib/postgresql.py:34]: I would love for this to be optional, in case the user wants to manage the schema manually. I could also see edge cases happening where, for some reason, we try to initialize this before postgres is available.

File: flipper/contrib/postgresql.py
Comment (adambom) [reply flipper/contrib/postgresql.py:48]: Any reason you did not use a separate (bigint or uuid) primary key?

File: tests/contrib/test_postgresql.py
Comment (adambom) [inline tests/contrib/test_postgresql.py:18]: These tests are rather slooow. I don't know that I care _that_ much because the project is small enough that it isn't a big deal. But... if there's anything you can do to make these run faster, it would be much appreciated. Thanks.

File: setup.py
Comment (adambom) [inline setup.py:10]: I am a little worried about dependency version conflicts here. It's also a bit of a weighty dependency for someone who doesn't want to use it.

The easy workaround is just to add a new `extras_require` section for "postgres, i.e. `pip install flipper-client[postgres]`. That allows users to opt-in if they want to use it. Honestly, we should probably do this for all the backends we support.

**Review by adambom — COMMENTED** [review]
Terrific! Thanks for the contribution. Everything here looks like it's almost ready to merge. I've made some suggestions inline about ways I think we could improve even further. I don't think anything I've suggested will be a breaking change compared to what you already have,.

File: setup.py
Comment (manuelxarez) [reply setup.py:10]: Great that you touched this topic. I meant to get some feedback on this: 
`psycopg` requires `libpq` to be installed in the system. Because we import every backend to the namespace, a user who isn't using the postgreSQL backend will still need to have `libpq` installed otherwise the import of flipper fails.

I couldn't think of a way to fix this that doesn't break backwards compatibility

File: flipper/contrib/postgresql.py
Comment (manuelxarez) [reply flipper/contrib/postgresql.py:48]: No specific reason, would you prefer to have a separate primary key?

File: tests/contrib/test_postgresql.py
Comment (manuelxarez) [reply tests/contrib/test_postgresql.py:18]: I'll investigate :+1:

File: setup.py
Comment (adambom) [reply setup.py:10]: Yeah I think the easiest thing to do is to add a separate `postgres` target to the `extras_require` section. That forces the user to opt-in to installing postgres. Once opted-in, then we can add instructions to the readme around specific installation requirements for postgres.

Comment (adambom) [reply setup.py:10]: Oh, got it. Thanks for the clarification. I can think of two ways around this. Python (for better or worse) allows you to import dependencies inline (In a method, for example). You can always try/except too. I think this is my preferred approach:

```python
try:
  import psycopg
except ModuleNotFoundError:
  pass
```

Comment (adambom) [reply setup.py:10]: If you want to be extra careful, you could even set a module-level flag:

```python
POSTGRES_ENABLED = False
try:
  import psycopg
  POSTGRES_ENABLED = True
except ModuleNotFoundError:
  pass


class PostgreSQLFeatureFlagStore:
  # ...
  def __init__(self, *args, **kwargs):
    if POSTGRES_ENABLED is False:
       raise PostgresNotEnabled()
```

Comment (manuelxarez) [reply setup.py]: I used your approach but in the `contrib/__init__.py` file (I think its more extendable for the future with other optional stores), let me know what you think.

I really appreciate the time your taking to give feedback :bow:

File: tests/contrib/test_postgresql.py
Comment (manuelxarez) [reply tests/contrib/test_postgresql.py:18]: I couldn't figure out a way to speed the tests here :(

File: flipper/__init__.py
Comment (adambom) [inline flipper/__init__.py:17]: Our convention is to respect rule 403 when at all possible. Could you please go back to the explicit imports?

File: flipper/contrib/__init__.py
Comment (adambom) [inline flipper/contrib/__init__.py:37]: I'd prefer to do the conditional import in the `PostgreSQLFeatureFlagStore` so we don't have to do the import *.

**Comment by manuelxarez** [comment]
@adambom Addressed your comments, let me know :+1:

**Comment by manuelxarez** [comment]
> This looks terrific. I'm going to merge and release. Please make sure to [sign the CLA](https://www.clahub.com/agreements/carta/flipper-client).

Hey @adambom it seems CLAHub is down (https://github.com/clahub/clahub/issues/174), at least cannot access it on my side.

**Comment by manuelxarez** [comment]
HI @adambom, any news on this? :bow:

**Comment by adambom** [comment]
> HI @adambom, any news on this? 🙇

Sorry for the delay. We're working on getting this merged. I can't approve since technically I opened the PR, and I can't merge it without an approval. Then this just fell through the cracks. From my end it's LGTM.

**Review by CaleMacDonald — APPROVED** [review]

**Comment by adambom** [comment]
Gonna release this today

**Comment by adambom** [comment]
Published as 1.3.0: https://pypi.org/project/flipper-client/1.3.0/

**Pull request merged**
